src/pybind/mgr/dashboard/controllers/home.py

```diff
@@ -16,52 +16,79 @@
 from .. import mgr, logger
 
 
-LANGUAGES = {f for f in os.listdir(mgr.get_frontend_path())
-             if os.path.isdir(os.path.join(mgr.get_frontend_path(), f))}
-LANGUAGES_PATH_MAP = {f.lower(): {'lang': f, 'path': os.path.join(mgr.get_frontend_path(), f)}
-                      for f in LANGUAGES}
-# pre-populating with the primary language subtag
-for _lang in list(LANGUAGES_PATH_MAP.keys()):
-    if '-' in _lang:
-        LANGUAGES_PATH_MAP[_lang.split('-')[0]] = {
-            'lang': LANGUAGES_PATH_MAP[_lang]['lang'], 'path': LANGUAGES_PATH_MAP[_lang]['path']}
-
+class LanguageMixin(object):
+    @lru_cache()
+    def languages(self):
+        # type: () -> set
+        return {
+            f
+            for f in os.listdir(mgr.get_frontend_path())
+            if os.path.isdir(os.path.join(mgr.get_frontend_path(), f))
+        }
 
-def _get_default_language():
-    with open("{}/../package.json".format(mgr.get_frontend_path()), "r") as f:
-        config = json.load(f)
-    return config['config']['locale']
+    @lru_cache()
+    def languages_path_map(self):
+        # type: () -> dict
+        languages_path_map = {
+            f.lower(): {
+                'lang': f,
+                'path': os.path.join(mgr.get_frontend_path(), f)
+            }
+            for f in self.languages()
+        }
+        # pre-populating with the primary language subtag
+        for lang in list(languages_path_map.keys()):
+            if '-' in lang:
+                languages_path_map[lang.split('-')[0]] = {
+                    'lang': languages_path_map[lang]['lang'],
+                    'path': languages_path_map[lang]['path']
+                }
+        return languages_path_map
 
+    @lru_cache()
+    def default_language(self):
+        # type: () -> str
+        with open("{}/../package.json".format(mgr.get_frontend_path()),
+                  "r") as f:
+            config = json.load(f)
+        return config['config']['locale']
 
-DEFAULT_LANGUAGE = _get_default_language()
-DEFAULT_LANGUAGE_PATH = os.path.join(mgr.get_frontend_path(), DEFAULT_LANGUAGE)
+    @lru_cache()
+    def default_language_path(self):
+        # type: () -> str
+        return os.path.join(mgr.get_frontend_path(), self.default_language())
```

Comment:
The problem with the old implementation is that it's absolutely not patchable in unit tests. I used `lru_cache`because in Python2 there is not `cached_property` and i do not want to backport it. Second, the whole initialization of the locale paths must be done only once at startup, the path should never change during runtime; to implement that caching was the easiest and cheapest way.

Reply:
Ok, it's not patcheable because these were global constants, but what if you move those constants inside the class `__init__`? That won't be evaluated as long as the class is not instantiated:
```py
class HomeController(BaseController, LanguageMixin):
  ...
  def __init__(self, *args, **kwargs):
    super(...)
    self.LANGUAGES = {f for f in os.listdir(mgr.get_frontend_path())
             if os.path.isdir(os.path.join(mgr.get_frontend_path(), f))}
   ...
```
My point is that given this data is immutable, it doesn't make sense to make it mutable and then adding a cache layer to make it look immutable again.